### DISCUSSION TITLE
How to send KNX messages with docker on Synology

### DISCUSSION THREAD

**Discussion by mivola** [john30/ebusd#844] [Q&A] [answered]
Hi,

I have an old version of ebusd running on a Raspi. There I'm using a custom solution to send the ebus data to my KNX bus (and recieve data from KNX to change ebus values). 

Currently I'm playing around with ebusd in a Docker container on my Synology NAS and it is really exciting to see how easy it is to setup in general. So ebusd starts correctly, can read values and send the data to my MQTT broker. What I'm struggeling with is the direct KNX integration. I mapped a file from my Synology to /etc/ebusd/knx.cfg (and adjusted the "address" and GAs) and I added an environment variable "EBUSD_KNXURL" which maps to my KNX interface ""ip:wiregate:6720".

In the log I can see two entries rgd "knx":
```
2023-02-24 13:09:13.070 [knx notice] address not assigned yet, entering programming mode
2023-02-24 13:09:13.071 [knx notice] connected to KNXnet/IP multicast 
```
Nothing else can be found in the log and no messages appear on the KNX bus. Is there anything that I'm missing? 

Thanks a lot!
Michael

**Comment by mivola** [comment]
@john30 any idea where/how I could start investigating? I also had a look at the source code and found this:  https://github.com/john30/ebusd/blob/b5bedc9080433ba3dc172f172de63c724ed95159/src/ebusd/knxhandler.cpp#L766 it seems that right after this log message, some data should be sent to the KNX bus, but I dont see anything coming. 


I also checked this thread: https://knx-user-forum.de/forum/%C3%B6ffentlicher-bereich/knx-eib-forum/842619-neue-infos-zur-vaillant-knx-anbindung/page19 but since I'm using a Wiregate server to connect to the bus, I think I dont need knxd in the ebusd docker image, right?

In the meantime I also managed to get Node Red up & running in a Docker container. In there I'm using knxEasy and as a connection I specified: "Wiregate-IP:3671" and it works! So this is using multicast, correct? But using the same ip/port and even "EBUSD_KNXURL": "@Wiregate-IP" (as suggested [here](https://knx-user-forum.de/forum/%C3%B6ffentlicher-bereich/knx-eib-forum/842619-neue-infos-zur-vaillant-knx-anbindung?p=1799076#post1799076)) doesnt work for ebusd.

Thanks,
Michael

**Reply by XueSheng-GIT** [reply]
I'm struggling with same same issue. I'm trying to get ebusd docker running. Reading the values via ebusctl is no issue. But the communication via knx doesn't seem to work. I'm using a wiregate as KNX device (routing enabled).
Docker run command:
```
docker run -d --name=ebusd -v /etc/ebusd:/etc/ebusd --device=/dev/ebus-usb:/dev/ttyUSB0 -p 8888 -e "EBUSD_KNXURL=@WIREGATE-IP" -e "EBUSD_KNXINT=/etc/ebusd/knx.cfg" john30/ebusd
```

Any further hint how to solve this issue?

**Comment by XueSheng-GIT** [comment]
After thinking about it, it seems to be a general networking issue. Docker container does not receive the multicast. Workaround could be using host network or macvlan. I'm not an networking expert, but this seems makes things more complicated. Because "--net=host" didn't do the trick in the first step, I just compiled ebusd myself and started it locally.
After using "--knxurl=" it seems to work.

Imho the docker image should include knxd by default, so that tunneling could be used. Otherwise it seems that it's getting quite complicated in regards to the networking configuration.

**Reply by c-huetter** [reply]
FYI: I have created a [pull request](https://github.com/john30/ebusd/pull/1051) that adds a build target for ebusd docker containers with built-in knxd support.

**Reply by mivola** [reply]
Sounds interesting. How can I create this container locally so I could test it? And maybe even let me know if you can help me with this one?

**Reply by c-huetter** [reply]
Creating an ebusd container with knxd support is quite easy now:
```shell
git clone https://github.com/c-huetter/ebusd-with-knxd.git
cd ebusd-with-knxd/contrib/docker
./build.sh knxd
docker run john30/ebusd:knxd
``` 
Happy to help if you have further questions.

**Reply by mivola** [reply]
I followed your steps but it fails on my Synology:
```
micha@mediaserver:~/_git/ebusd-with-knxd/contrib/docker$ ll
total 48
drwxrwxrwx+ 1 micha users  314 Nov  2 18:22 .
drwxrwxrwx+ 1 micha users  138 Nov  2 18:22 ..
-rwxrwxrwx+ 1 micha users 2379 Nov  2 18:22 build.sh
-rwxrwxrwx+ 1 micha users 7364 Nov  2 18:22 docker-compose.example.yaml
-rwxrwxrwx+ 1 micha users   82 Nov  2 18:22 docker-entrypoint.sh
-rwxrwxrwx+ 1 micha users 1312 Nov  2 18:22 Dockerfile
-rwxrwxrwx+ 1 micha users 1343 Nov  2 18:22 Dockerfile.knxd
-rwxrwxrwx+ 1 micha users 1379 Nov  2 18:22 Dockerfile.release
-rwxrwxrwx+ 1 micha users 1220 Nov  2 18:22 Dockerfile.template
drwxrwxrwx+ 1 micha users   36 Nov  2 18:22 hooks
-rwxrwxrwx+ 1 micha users  410 Nov  2 18:22 manifest.template
-rwxrwxrwx+ 1 micha users 4525 Nov  2 18:22 README.md
-rwxrwxrwx+ 1 micha users 2395 Nov  2 18:22 update.sh
micha@mediaserver:~/_git/ebusd-with-knxd/contrib/docker$ docker -v
Docker version 20.10.23, build 876964a
micha@mediaserver:~/_git/ebusd-with-knxd/contrib/docker$ ./build.sh knxd
unknown flag: --target
See 'docker --help'.
```
Now I'm stuck again :-(

**Reply by c-huetter** [reply]
It seems like the docker features differ significantly between DSM releases: My docker was complaining that it doesn't support the [buildx](https://docs.docker.com/engine/reference/commandline/buildx/) plugin, yours doesn't know the [--target](https://docs.docker.com/engine/reference/commandline/build/#target) option. The following command has worked for me:
```shell
docker build \
    --target image \
    --progress plain \
    --platform linux/amd64 \
    -f Dockerfile.knxd \
    --build-arg "BASE_IMAGE=debian:bullseye" \
    --build-arg "EBUSD_VERSION=23.2" \
    --build-arg "EBUSD_IMAGE=bullseye" \
    --build-arg "TARGETARCH=amd64" \
    -t john30/ebusd:knxd \
    .
```

**Reply by mivola** [reply]
That works a bit better, but still throwing an error in step 13/30:
```
Step 12/30 : ADD . /build
 ---> e82b83407b82
Step 13/30 : RUN RUNTEST=full GIT_REVISION=$GIT_REVISION ./make_debian.sh --with-knxd
 ---> Running in c0fb56def37d
/bin/sh: 1: ./make_debian.sh: not found
The command '/bin/sh -c RUNTEST=full GIT_REVISION=$GIT_REVISION ./make_debian.sh --with-knxd' returned a non-zero code: 127

```

**Comment by mivola** [comment, marked as answer]
OK, I think we need to go one step back. After a lot of tests it seems the knx.cfg is not even used by the ebusd in the Docker container. Even if I have an invalid "address" in there, the knx related log messages are the same - no error message!
But using "EBUSD_KNXVAR"="address=1.x1.y111" (as described [here](https://github.com/john30/ebusd/wiki/2.-Run#knx-options)), does lead to an error in the log:
```
[knx error] invalid address: 1.x1.y111
```

But even with a correct address, I didnt see any KNX messages on the bus, but the direction seems to be right ...

Next I added "EBUSD_KNXINT"="/etc/ebusd/knx.cfg" and defined an invalid address there as well - and this leads to the same error message as using "EBUSD_KNXVAR"! Now I was sure, that the knx.cfg was actually used!

Next was about the network. I played a lot with "EBUSD_KNXURL" & the Docker network setup. Eventually I got it working with the following setup in the host network:

![image](https://user-images.githubusercontent.com/743342/225112359-4be7fd03-a5bd-42d3-968a-bf5b6e67bdab.png)
![image](https://user-images.githubusercontent.com/743342/225112430-75cadea8-d2ac-4b97-9cb9-53c54fff064c.png)

@XueSheng-GIT maybe you can give it a try and test this setup? 

If you can confirm my findings, it seems the root cause is that the default value for "EBUSD_KNXINT" is not "/etc/ebusd/knx.cfg" ...

**Reply by XueSheng-GIT** [reply]
Thanks @mivola, good catch. It seems to make a difference if EBUSD_KNXINT is used. But as soon as I enable --net=host, it seems knxd gets blocked by ebusd (e.g. I cannot access knxd by ETS anymore).
```
$ docker run -d --name=ebusd --restart unless-stopped --net=host -v /etc/ebusd:/etc/ebusd --device=/dev/ttyEBUS:/dev/ttyUSB0 -e EBUSD_KNXINT="/etc/ebusd/knx.cfg" -e EBUSD_KNXURL="" john30/ebusd
```

I'll not invest any more time in this issue. I'm already back to yawgd to connect ebusd docker to knx. This works without any hickups (read and write to knx and I get my RRDs updated). When using ebusd locally (without docker) there where also a couple of issues (some values could not be read from the bus, writing also wasn't straight forward, I saw also a couple of other reports that negative temperature values are not being sent correctly.

**Reply by mivola** [reply]
> I'm already back to yawgd to connect ebusd docker to knx. 

Can you maybe explain a little more details how you setup the yawgd and the telnet connection to ebusd in Docker?

**Reply by XueSheng-GIT** [reply]
> Can you maybe explain a little more details how you setup the yawgd and the telnet connection to ebusd in Docker?

Basically I did the following steps (short form, let me know if you need further details on some of the steps).
- ebusd is running in docker, yawgd and knxd are installed on the host machine itself!
- add `-p 127.0.0.1:8888:8888` to your ebusd docker run (makes telnet port 8888 available on localhost)
- I've updated yawgd to be compatible with latest ebusd/knxd (see https://github.com/XueSheng-GIT/yawgd). There are still some parts which I may rewrite if I find the time, but generally this is working as it is right now).
- Create a dump of the ebusd config (some hints in this discussion: https://github.com/john30/ebusd/discussions/720#discussioncomment-5319375)
- Use `/etc/yawgd/tools/make_ebus_config.pl` to create the required yawgd-ebusd.csv
- add your GAs to yawgd-ebusd.csv
- after restarting yawgd, double check `/var/log/yawgd-plugin.log` for errors.

**Reply by mivola** [reply]
> > Can you maybe explain a little more details how you setup the yawgd and the telnet connection to ebusd in Docker?
> 
> Basically I did the following steps (short form, let me know if you need further details on some of the steps).
> 
> * ebusd is running in docker, yawgd and knxd are installed on the host machine itself!
> * I've updated yawgd to be compatible with latest ebusd/knxd (see https://github.com/XueSheng-GIT/yawgd). There are still some parts which I may rewrite if I find the time, but generally this is working as it is right now).

I'm still struggling with ebusd and gettings its value to knx and back using MQTT from the ebusd Docker container... 
Since I'm running the Docker on my Synology, I'd like to _not_ install yawgd&knxd onto the host - do you think it would be (easily) possible to get yawgd (and knxd?) into a Docker container and let it talk from there to the ebusd in another Docker container?

**Reply by XueSheng-GIT** [reply]
> Since I'm running the Docker on my Synology, I'd like to not install yawgd&knxd onto the host - do you think it would be (easily) possible to get yawgd (and knxd?) into a Docker container and let it talk from there to the ebusd in another Docker container?

I'm sorry, but Docker is still some kind of black box to me. Thus, I don't know whether it would be easy or not to create a knxd/yawgd docker.

**Comment by c-huetter** [comment]
> I'm still struggling with ebusd and gettings its value to knx and back using MQTT from the ebusd Docker container... Since I'm running the Docker on my Synology, I'd like to _not_ install yawgd&knxd onto the host - do you think it would be (easily) possible to get yawgd (and knxd?) into a Docker container and let it talk from there to the ebusd in another Docker container?

I only got the ebusd container to work on my Synology Diskstation through a knxd container. The reason is that Synology doesn't forward multicast IP traffic from containers, even if you use the `host` network setting. The [knxd container from renehezser](https://registry.hub.docker.com/r/renehezser/knxd/) works quite nicely on my Diskstation. Just let me know if you need help with configuring knxd or ebusd.

**Reply by mivola** [reply]
Well, I'm just not sure which way to go now. Basically I'd love to have ebusd in a Docker container on my Synology talking to KNX via my WireGate. 
But should I use your PR to create a [combined container with ebusd & knxd](https://github.com/john30/ebusd/discussions/844#discussioncomment-7421803)? Or use the [knxd container from renehezser](https://registry.hub.docker.com/r/renehezser/knxd/) that you mentioned? 
Any suggestion & guide how to configure would be appreciated!
Thanks,
Michael

**Reply by c-huetter** [reply]
Hi Michael,

Since I didn't manage to connect Docker containers to my KNX IP router, I would recommend to deploy two separate containers:

1. ready-made knxd container (e.g. [from renehezser](https://registry.hub.docker.com/r/renehezser/knxd/)) which connects to your KNX gateway through an IP tunnel (`knxd -b ipt:gateway-ip-address`)
2. custom-build ebusd container (see my [pull request](https://github.com/john30/ebusd/pull/1051)) that connects to your knxd container (`ebusd --knxurl=ip:localhost`)

Best,
Christian

**Reply by mivola** [reply]
> 1. ready-made knxd container (e.g. [from renehezser](https://registry.hub.docker.com/r/renehezser/knxd/)) which connects to your KNX gateway through an IP tunnel (`knxd -b ipt:gateway-ip-address`)
> 2. custom-build ebusd container (see my [pull request](https://github.com/john30/ebusd/pull/1051)) that connects to your knxd container (`ebusd --knxurl=ip:localhost`)

Hi Christian,
that sounds promising, I'll give it a try. Is that the setup you're successfully using?
But how is the mapping between ebusd messages & KNX GAs done? Via the [knx.cfg in the ebus container](https://github.com/john30/ebusd/blob/master/contrib/etc/ebusd/knx.cfg)? If yes, the problem I described [here](https://github.com/john30/ebusd/discussions/855) is not related to an incorrect mapping but to the not-forwarding of multicast traffic of containers?
Thanks a lot!
Michael

**Reply by mivola** [reply]
I think I need to investigate more into the knx.cfg mapping, b/c after having the container running over night (with --host as described [here](https://github.com/john30/ebusd/discussions/844#discussioncomment-5313732)), I indeed found a few messages in my eib.log on the WireGate. So I played a bit more with the knx.cfg and now I successfully get messages from ebusd to knx w/o a dedicated knxd:
```
2023-11-03 07:09:06.401,A_GroupValue_Write,1.1.253,4/0/1,05 D4,,,,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:09:06.422,A_GroupValue_Write,1.1.253,4/0/0,01,,,,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:09:06.456,A_GroupValue_Write,1.1.253,4/0/2,00,,,,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:09:18.474,A_GroupValue_Write,1.1.253,4/0/5,01,,,,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:09:23.257,A_GroupValue_Write,1.1.253,4/0/3,00 00 00 11,,,,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:09:23.290,A_GroupValue_Write,1.1.253,4/0/2,01,,,,0,low,5,T_DATA_XXX_REQ,0

2023-11-03 07:16:09.570,A_GroupValue_Write,1.1.253,4/1/15,05 5F,13.75,,9,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:16:21.903,A_GroupValue_Write,1.1.253,4/1/25,6C,83886.08,,9,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:16:52.226,A_GroupValue_Write,1.1.253,4/1/15,05 58,13.68,,9,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:17:12.373,A_GroupValue_Write,1.1.253,4/1/25,6B,62914.56,,9,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:17:21.640,A_GroupValue_Write,1.1.253,4/1/12,06 1A,15.62,,9,0,low,5,T_DATA_XXX_REQ,0
2023-11-03 07:17:52.720,A_GroupValue_Write,1.1.253,4/1/15,05 52,13.62,,9,0,low,5,T_DATA_XXX_REQ,0
```
Here is the knx.cfg setting for that:
```
global/running = 4/0/0
global/version = 4/0/1
global/signal = 4/0/2
global/uptime = 4/0/3
global/updatecheck = 4/0/4
global/scan = 4/0/5

broadcast/datetime/outsidetemp = 4/1/10
ehp/TempInput/temp = 4/1/12
ehp/SourceTempInput/temp = 4/1/15
ehp/ActualEnvironmentPower/power = 4/1/23
ehp/ActualEnvironmentPowerFine/0 = 4/1/25
```

So maybe your assumption that Synology doesnt forward the multicast traffic is not (always) correct or my setup is a bit different?

**Reply by c-huetter** [reply]
That's really interesting, maybe I should give multicasting another try!

Regarding the mapping of ebus fields to knx group addresses: the format "circuit/message/field name" with three components is hard-coded in [knxhandler.cpp](https://github.com/john30/ebusd/blob/3d9dd9b1c3f9cc7e9f59c177883dd906c5294e5d/src/ebusd/knxhandler.cpp#L828), only global fields may use two components. Make sure that your configuration csv file has all three compontents filled. `ebusctl` can handle messages with empty field names, but knxhandler will ignore them.
